File: collision_avoidance_real_torque_playback/logs/plot_dq.py
import xml.etree.ElementTree as ET
import matplotlib.pyplot as plt
import numpy as np

filename = 'dq_robot.xml'

tree = ET.parse(filename)
root = tree.getroot()

keypoints = root.findall('.//keypoint')
N = len(keypoints)

time = np.zeros(N)
q = np.zeros((N, 7))  # One column for each point id

for i, kp in enumerate(keypoints):
    # Time
    time[i] = float(kp.attrib['time'])

    # Seven point values
    pts = kp.findall('point')
    for j in range(7):
        q[i, j] = float(pts[j].text)

# Plot all seven trajectories
plt.figure()
t = np.linspace(0, (N - 1) / 1000, N)
plt.plot(t, q)

plt.grid(True)
plt.xlabel('Time (s)')
plt.ylabel('Joint speed')
plt.title('Robot joint speed')

# Commented legend
plt.legend([f'Joint {i}' for i in range(7)], loc='best')

plt.show()

File: collision_avoidance_real_torque_playback/logs/plot_dq.m
filename = 'dq_robot.xml';

doc = xmlread(filename);
keypoints = doc.getElementsByTagName('keypoint');

N = keypoints.getLength;

time = zeros(N,1);
q = zeros(N,7);    % one column for each point id

for i = 0:N-1
    kp = keypoints.item(i);

    % Time
    time(i+1) = str2double(char(kp.getAttribute('time')));

    % Seven point values
    pts = kp.getElementsByTagName('point');
    for j = 0:6
        q(i+1,j+1) = str2double(char(pts.item(j).getTextContent));
    end
end

% Plot all seven trajectories
figure
t = 0:.001:(N-1)/1000;
plot(t', q, 'LineWidth', 1.5)

grid on
xlabel('Time (s)')
ylabel('Joint speed')
title('Robot joint speed')

legend('Joint 0','Joint 1','Joint 2','Joint 3', ...
    'Joint 4','Joint 5','Joint 6', ...
    'Location','best')

hold on;
% vertical lines
% vlineTimes  = [3, 4, 7, 9, 11];
% vlineColor  = [0.4 0.4 0.4];   % dark grey
% vlineStyle  = '--';
% vlineWidth  = 1.2;
% 
% hVline = gobjects(1, numel(vlineTimes));
% for i = 1:numel(vlineTimes)
%     hVline(i) = xline(vlineTimes(i), ...
%         vlineStyle, ...
%         'Color',     vlineColor, ...
%         'LineWidth', vlineWidth, ...
%         'DisplayName', sprintf('t = %g', vlineTimes(i)));
% end